CoreScriptsRoot/Modules/VR/Dialog.lua

```diff
@@ -0,0 +1,220 @@
+local DIALOG_BG_TRANSPARENCY = 0.3
+
+local TITLE_COLOR = Color3.new(1, 1, 1)
+
+local DIALOG_TITLE_HEIGHT = 66
+local DIALOG_COLOR_HEIGHT = 8
+local DIALOG_TITLE_TEXT_SIZE = Enum.FontSize.Size36
+
+local DIALOG_CONTENT_PADDING = 48
+
+local emptySelectionImage = Utility:Create "ImageLabel" { Image = "", BackgroundTransparency = 1, ImageTransparency = 1 }
+
+local PANEL_OFFSET_CF = CFrame.new(0, 0, -7)* CFrame.Angles(0, math.pi, 0)
+
+local DialogPanel = Panel3D.Get("Dialog")
+DialogPanel:SetType(Panel3D.Type.Fixed)
+DialogPanel.localCF = PANEL_OFFSET_CF
+DialogPanel:SetCanFade(true)
+local dialogContentFrame = Utility:Create "Frame" {
+	Parent = DialogPanel:GetGUI(),
+	BackgroundTransparency = 1,
+	Position = UDim2.new(0, 0, 0, 0),
+	Size = UDim2.new(1, 0, 1, 0)
+}
+
+local dialogPanelAngle = 0
+
+local PANEL_FADE_ANGLE_0, PANEL_FADE_ANGLE_1 = math.rad(30), math.rad(40)
+local PANEL_FADE_RANGE = PANEL_FADE_ANGLE_1 - PANEL_FADE_ANGLE_0
+local PANEL_REAPPEAR_ANGLE = math.rad(90)
+
+local function positionDialogPanel(desiredAngle)
+	dialogPanelAngle = desiredAngle
+	local headCF = InputService:GetUserCFrame(Enum.UserCFrame.Head)
+	local headPos = headCF.p
+	DialogPanel.localCF = CFrame.new(headPos) * CFrame.Angles(0, desiredAngle, 0) * PANEL_OFFSET_CF
+end
+
+function DialogPanel:CalculateTransparency()
+	local headCF = InputService:GetUserCFrame(Enum.UserCFrame.Head)
+	local headLook = headCF.lookVector * Vector3.new(1, 0, 1)
+	local vectorToPanel = Vector3.new(math.cos(dialogPanelAngle + math.rad(90)), 0, -math.sin(dialogPanelAngle + math.rad(90)))
+
+	local angleToPanel = math.acos(headLook:Dot(vectorToPanel))
+	return math.min(math.max(0, (angleToPanel - PANEL_FADE_ANGLE_0) / PANEL_FADE_RANGE), 1)
+end
+
+
+local opacityLookup = {}
+
+game:GetService("RunService"):BindToRenderStep("DialogPanel", Enum.RenderPriority.Last.Value, function()
+	if DialogPanel.transparency == 1 then
+		local headCF = InputService:GetUserCFrame(Enum.UserCFrame.Head)
+		local headLook = headCF.lookVector * Vector3.new(1, 0, 1)
+		local headAngle = (math.atan2(-headLook.Z, headLook.X) - math.rad(90)) % math.rad(360)
+		local newPanelAngle = math.floor((headAngle / PANEL_REAPPEAR_ANGLE) + 0.5) * PANEL_REAPPEAR_ANGLE
+		positionDialogPanel(newPanelAngle)
+	end
+
+	--update the transparency of gui elements
+	local opacityMult = 1 - DialogPanel.transparency
+	for guiElement, baseOpacity in pairs(opacityLookup) do
+		local transparency = 1 - (baseOpacity * opacityMult)
+		if guiElement:IsA("TextLabel") or guiElement:IsA("TextButton") then
+			guiElement.TextTransparency = transparency
+		elseif guiElement:IsA("ImageLabel") or guiElement:IsA("ImageButton") then
+			guiElement.ImageTransparency = transparency
+		end
+	end
+end)
+
+local DialogQueue = {}
+
+local currentDescendantConn = nil
+local function updatePanel()
+	local currentDialog = DialogQueue[1]
+	if not currentDialog or not currentDialog.content then
+		dialogContentFrame:ClearAllChildren()
```

Comment:
I'd rather just recreate it when it's needed, since we won't get too many dialogs in a single play-session. This way, we shouldn't have to worry about the dialog frame state getting messed up by something externally.